File: back-end/src/main/java/nl/hva/backend/domain/models/game/board/location/locationImpl/Utility.java
package nl.hva.backend.domain.models.game.board.location.locationImpl;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.board.GameBoard;
import nl.hva.backend.domain.models.game.board.location.OwnableLocation;

import javax.persistence.Entity;

@Entity
@NoArgsConstructor
@Getter
@Setter
public class Utility extends OwnableLocation {
    private String name;
    private double diceRollRentFactor;
    private double diceRollRentFactorIfTwoAreOwned;

    public Utility(
            long id,
            int position,
            String description,
            GameBoard gameBoard,
            double initialPurchasePrice,
            String name,
            double diceRollRentFactor,
            double diceRollRentFactorIfTwoAreOwned
    ) {
        super(id, position, description, gameBoard, initialPurchasePrice);
        this.name = name;
        this.diceRollRentFactor = diceRollRentFactor;
        this.diceRollRentFactorIfTwoAreOwned = diceRollRentFactorIfTwoAreOwned;
    }
}

File: front-end/src/app/domain/models/game/board/location/locationImpl/Utility.ts
import {OwnableLocation} from "../OwnableLocation";

export class Utility extends OwnableLocation {
  private _name: string;
  private _diceRollRentFactor: number;
  private _diceRollRentFactorIfTwoAreOwned: number;

  public constructor(
    id: number,
    position: number,
    description: string,
    initialPurchasePrice: number,
    name: string,
    mortgageValue: number,
    diceRollRentFactor: number,
    diceRollRentFactorIfTwoAreOwned: number,
  ) {
    super(Utility.name, id, position, description, initialPurchasePrice, mortgageValue);
    this._name = name;
    this._diceRollRentFactor = diceRollRentFactor;
    this._diceRollRentFactorIfTwoAreOwned = diceRollRentFactorIfTwoAreOwned;
  }

  public static trueCopy(location: Utility): Utility | null {
    return new Utility(
      location.id,
      location.position,
      location.description,
      location.initialPurchasePrice,
      location.name,
      location.mortgageValue,
      location.diceRollRentFactor,
      location.diceRollRentFactorIfTwoAreOwned,
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(),
      {
        "initialPurchasePrice": this.initialPurchasePrice,
        "name": this.name,
        "mortgageValue": this.mortgageValue,
        "diceRollRentFactor": this.diceRollRentFactor,
        "diceRollRentFactorIfTwoAreOwned": this.diceRollRentFactorIfTwoAreOwned,
      }
    );
  }

  public get name(): string {
    return this._name;
  }

  public get diceRollRentFactor(): number {
    return this._diceRollRentFactor;
  }

  public get diceRollRentFactorIfTwoAreOwned(): number {
    return this._diceRollRentFactorIfTwoAreOwned;
  }

  public override toString(): string {
    return this.name;
  }

}
